### ISSUE TITLE
NCCL is not supported when using virtual GPUs, fallingback to reduction to one device

### ISSUE THREAD

**Issue by asm582** [tensorflow/tensorflow#40735]
Hello,

I tried to run colab notebook present here: https://www.tensorflow.org/guide/gpu.

When I try to use virtual GPU I get NCCL not supported error. can you help on how do I test virtual GPU's in colab?

**Comment by Saduf2019** [comment]
@asm582 
Please select "Runtime tab" and option "change runtime" in this option select "GPU".
to confirm, run:

'''
_lmport tensorflow as tf
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))_ '''

Can you please refer to [this link](https://medium.com/deep-learning-turkey/google-colab-free-gpu-tutorial-e113627b9f5d) and let me know if it helps resolve your issue.
[link](https://hackernoon.com/begin-your-deep-learning-project-for-free-free-gpu-processing-free-storage-free-easy-upload-b4dba18abebc). [link2](https://jovianlin.io/pytorch-with-gpu-in-google-colab/)

**Comment by nikitamaia** [comment]
Hi @asm582, was that an error message or a warning? From the message you've shared, I would guess that you're trying to run multiworker training on virtual GPUs in colab. In this case, NCCL is not supported, so as the message suggests [tf.distribute.ReductionToOneDevice](https://www.tensorflow.org/api_docs/python/tf/distribute/ReductionToOneDevice) will be used instead to aggregate the gradients. Happy to help more if you provide additional information on what you're trying to do and where specifically you saw this message.

**Comment by Antymon** [comment]
@nikitamaia What does it mean for GPU to be virtual and how to make it non-virtual? I have access to the bare metal of a multi-GPU node, operate on "PhysicalDevices" allocating one process per GPU and keep seeing the same warning message.

**Comment by nikitamaia** [comment]
More on virtual GPUs here https://www.tensorflow.org/guide/gpu#manual_device_placement
Feel free to open a new issue if more support is needed.

**Comment by Antymon** [comment]
I am still strictly after the warning mentioned in this issue. Docs you pointed to seem to suggest that one needs to explicitly setup Virtual GPUs for them to pop up, whereas I have never done such a thing and still seeing "NCCL is not supported when using virtual GPUs..." when using simplest data paralllelism.

**Comment by animesh0794** [comment]
@nikitamaia is tf.distribute.ReductionToOneDevice less efficient than NCCL? Or what exactly are the differences that using these two reduction methods bring on performance?

**Comment by nikitamaia** [comment]
@animesh0794, if you have NCCL on your system then that is the best option for performance. `ReductionToOneDevice` is a good option when you do not have NCCL on your system.

**Comment by animesh0794** [comment]
Hey Nikita,

Thanks, for replying!
I am using AWS Sagemaker and I am not sure how can I check if NCCL is
installed on that virtual machine.

Thanks,
Animesh
